activerecord/lib/active_record/autosave_association.rb

```diff
@@ -412,6 +414,32 @@ def save_collection_association(reflection)
         end
       end
 
+      # In the case where both ends of the relation have autosave: true, we can
+      # end up in a scenario where the save on the parent triggers a save on the
+      # child, triggering a second, duplicate save on the parent again.
+      #
+      # In order to avoid this we temporarily disable autosave on associations
+      # leading back from child records to this parent.
+      def with_disabled_autosave_on_inverses(klass)
+        inverse_reflections_with_autosave = Set.new
```

Comment:
If you are only using this for linear `each`, what's the reason for using `Set` rather than normal `[]`?